File: tests/run_shared.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Runner de la batería compartida (tests/fixtures/fixtures.json) contra dedup.py.
La implementación JS (fase 2) debe pasar este MISMO json con su propio runner.

Contrato de "expect" (cada fixture):
  - El dedup.ris se evalúa EN MINÚSCULAS: toda comparación de substrings es
    case-insensitive y los valores del json deben escribirse en minúsculas.
  - Claves obligatorias (counts):
      duplicados            nº de filas de duplicados.csv
      unicos                nº de líneas de dedup.ris que empiezan por "ER  -"
      dudosos               nº de filas de revisar.csv
  - Claves opcionales:
      ris_contiene          substrings que deben aparecer en dedup.ris
      ris_no_contiene       substrings que NO deben aparecer en dedup.ris
      ris_una_vez           substrings que aparecen exactamente 1 vez en dedup.ris
      motivos_incluyen      pertenencia EXACTA al conjunto de valores de la
                            columna "motivo" de duplicados.csv
      motivos_duda_incluyen substring sobre la columna "motivo_duda" de
                            revisar.csv (todas las filas concatenadas)
      dudosos_texto_incluye substring sobre las columnas "titulo_A"+"titulo_B"
                            de revisar.csv (todas las filas concatenadas,
                            en minúsculas)
  - Cualquier otra clave en expect es un error del fixture y hace fallar el test.
  - Un stderr no vacío de dedup.py (AVISO de export truncado, 0 registros...)
    cuenta como fallo del fixture: la batería es estática y no debe avisar.
"""
import os, sys, csv, json, shutil, subprocess, tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
SCRIPT = os.path.join(HERE, "..", "scripts", "dedup.py")

CLAVES_OBLIGATORIAS = {"duplicados", "unicos", "dudosos"}
CLAVES_CONOCIDAS = CLAVES_OBLIGATORIAS | {
    "ris_contiene", "ris_no_contiene", "ris_una_vez",
    "motivos_incluyen", "motivos_duda_incluyen", "dudosos_texto_incluye",
}

def ejecuta(fx):
    """Ejecuta dedup.py sobre el fixture y devuelve (ris, dups, rev, stderr).
    OJO: el RIS se devuelve pasado a minúsculas (comparaciones case-insensitive)."""
    out = tempfile.mkdtemp(prefix="dedup_fx_")
    try:
        files = [os.path.join(HERE, f) for f in fx["files"]]
        proc = subprocess.run([sys.executable, SCRIPT] + files +
                              ["--source-names", fx["source_names"], "--out", out],
                              check=True, capture_output=True, text=True)
        with open(os.path.join(out, "dedup.ris"), encoding="utf-8") as f:
            ris = f.read().lower()
        with open(os.path.join(out, "duplicados.csv"), encoding="utf-8") as f:
            dups = list(csv.DictReader(f))
        with open(os.path.join(out, "revisar.csv"), encoding="utf-8") as f:
            rev = list(csv.DictReader(f))
        return ris, dups, rev, proc.stderr
    finally:
        shutil.rmtree(out, ignore_errors=True)

def cuenta_unicos(ris):
    return sum(1 for linea in ris.splitlines() if linea.startswith("er  -"))

def comprueba(fx):
    e, fallos = fx["expect"], []
    for k in sorted(set(e) - CLAVES_CONOCIDAS):
        fallos.append("clave expect desconocida: %s" % k)
    for k in sorted(CLAVES_OBLIGATORIAS - set(e)):
        fallos.append("falta clave expect obligatoria: %s" % k)
    if fallos:
        return fallos
    try:
        ris, dups, rev, stderr = ejecuta(fx)
    except subprocess.CalledProcessError as err:
        return ["dedup.py falló (exit %d): %s" % (err.returncode, (err.stderr or "").strip())]
    def chk(cond, msg):
        if not cond: fallos.append(msg)
    chk(not stderr, "stderr inesperado: %r" % stderr[:200])
    chk(len(dups) == e["duplicados"], "duplicados: %d != %d" % (len(dups), e["duplicados"]))
    chk(cuenta_unicos(ris) == e["unicos"], "unicos: %d != %d" % (cuenta_unicos(ris), e["unicos"]))
    chk(len(rev) == e["dudosos"], "dudosos: %d != %d" % (len(rev), e["dudosos"]))
    for s in e.get("ris_contiene", []): chk(s in ris, "falta en ris: %r" % s)
    for s in e.get("ris_no_contiene", []): chk(s not in ris, "sobra en ris: %r" % s)
    for s in e.get("ris_una_vez", []): chk(ris.count(s) == 1, "no exactamente 1 vez: %r" % s)
    mot = {d["motivo"] for d in dups}
    for s in e.get("motivos_incluyen", []): chk(s in mot, "falta motivo: %r" % s)
    duda = " | ".join(r["motivo_duda"] for r in rev)
    for s in e.get("motivos_duda_incluyen", []): chk(s in duda, "falta motivo_duda: %r" % s)
    texto = " ".join((r["titulo_A"] + r["titulo_B"]).lower() for r in rev)
    for s in e.get("dudosos_texto_incluye", []): chk(s in texto, "falta en dudosos: %r" % s)
    return fallos

def main():
    with open(os.path.join(HERE, "fixtures", "fixtures.json"), encoding="utf-8") as f:
        data = json.load(f)
    ok = True
    for fx in data["fixtures"]:
        fallos = comprueba(fx)
        print(("PASS" if not fallos else "FAIL"), "-", fx["name"])
        for msg in fallos:
            print("       ", msg); ok = False
    print("\n== TODOS EN VERDE ==" if ok else "\n== HAY FALLOS ==")
    sys.exit(0 if ok else 1)

if __name__ == "__main__":
    main()

File: tests/run_shared.mjs
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { deduplicar } from "../docs/engine/index.js";

const HERE = dirname(fileURLToPath(import.meta.url));
const CLAVES = new Set(["duplicados","unicos","dudosos","motivos_incluyen",
  "ris_contiene","ris_no_contiene","ris_una_vez","dudosos_texto_incluye","motivos_duda_incluyen"]);
const OBLIG = ["duplicados","unicos","dudosos"];

function comprueba(fx) {
  const e = fx.expect, fallos = [];
  for (const k of Object.keys(e)) if (!CLAVES.has(k)) fallos.push(`clave expect desconocida: ${k}`);
  for (const k of OBLIG) if (!(k in e)) fallos.push(`falta clave expect obligatoria: ${k}`);
  if (fallos.length) return fallos;

  let out;
  try {
    const entradas = fx.files.map((f, i) => ({
      name: f,
      text: readFileSync(join(HERE, f), "utf8"),
      source: fx.source_names.split(",")[i].trim(),
    }));
    out = deduplicar(entradas);
  } catch (err) { return [`motor lanzó: ${err.message}`]; }

  const ris = out.ris.toLowerCase();
  const chk = (c, m) => { if (!c) fallos.push(m); };
  const unicos = ris.split(/\r\n|\r|\n/).filter(l => l.startsWith("er  -")).length;
  chk(out.dups.length === e.duplicados, `duplicados: ${out.dups.length} != ${e.duplicados}`);
  chk(unicos === e.unicos, `unicos: ${unicos} != ${e.unicos}`);
  chk(out.review.length === e.dudosos, `dudosos: ${out.review.length} != ${e.dudosos}`);
  for (const s of e.ris_contiene ?? []) chk(ris.includes(s), `falta en ris: ${s}`);
  for (const s of e.ris_no_contiene ?? []) chk(!ris.includes(s), `sobra en ris: ${s}`);
  for (const s of e.ris_una_vez ?? []) chk(ris.split(s).length - 1 === 1, `no exactamente 1 vez: ${s}`);
  const mot = new Set(out.dups.map(d => d.motivo));
  for (const s of e.motivos_incluyen ?? []) chk(mot.has(s), `falta motivo: ${s}`);
  const duda = out.review.map(r => r.motivo_duda).join(" | ");
  for (const s of e.motivos_duda_incluyen ?? []) chk(duda.includes(s), `falta motivo_duda: ${s}`);
  const texto = out.review.map(r => (r.titulo_A + r.titulo_B).toLowerCase()).join(" ");
  for (const s of e.dudosos_texto_incluye ?? []) chk(texto.includes(s), `falta en dudosos: ${s}`);
  return fallos;
}

const data = JSON.parse(readFileSync(join(HERE, "fixtures", "fixtures.json"), "utf8"));
let ok = true;
for (const fx of data.fixtures) {
  const fallos = comprueba(fx);
  console.log((fallos.length ? "FAIL" : "PASS") + " - " + fx.name);
  for (const m of fallos) { console.log("        " + m); ok = false; }
}
console.log(ok ? "\n== TODOS EN VERDE ==" : "\n== HAY FALLOS ==");
process.exit(ok ? 0 : 1);
